lib/redis/connection/ruby.rb

```diff
@@ -34,20 +35,47 @@ def write_timeout=(timeout)
       end
 
       def read(nbytes)
-        result = @buffer.slice!(0, nbytes)
+        result = @buffer.byteslice(@offset, nbytes)
+        @offset += result.bytesize
 
         buffer = String.new(capacity: nbytes, encoding: Encoding::ASCII_8BIT)
```

Comment:
I also tested out removing this buffer and I actually get much better performance:

```
Calculating -------------------------------------
                Ruby     4.000  i/100ms
        Patched Ruby    19.000  i/100ms
-------------------------------------------------
                Ruby     52.216  (± 5.7%) i/s -    264.000
        Patched Ruby    196.500  (± 1.5%) i/s -    988.000

Comparison:
        Patched Ruby:      196.5 i/s
                Ruby:       52.2 i/s - 3.76x slower

Calculating -------------------------------------
                Ruby   127.698M memsize (     0.000  retained)
                        84.993k objects (     0.000  retained)
                        50.000  strings (     0.000  retained)
        Patched Ruby     1.208M memsize (     2.795k retained)
                        25.039k objects (     1.000  retained)
                        50.000  strings (     1.000  retained)

Comparison:
        Patched Ruby:    1208359 allocated
                Ruby:  127698147 allocated - 105.68x more
```

I understand that this was added due to https://github.com/redis/redis-rb/issues/935 though. Perhaps we could do this only in `SSLSocket`? This is not really related to the current patch though so we can discuss this separately.